File: lib/travis/synchronizer.rb
module Travis
  class Synchronizer < Array
    class << self
      def timeout
        @@timeout ||= 20
      end

      def timeout=(timeout)
        @@timeout = timeout
      end

      def synchronizers
        @@synchronizers ||= {}
      end

      def receive(id, msg_id, &forward)
        synchronizers[id] ||= new(id)
        synchronizers[id].receive(msg_id, &forward)
      end
    end

    attr_reader :id, :last_id, :finalizer

    def initialize(id)
      @id = id
      @last_id  = 0
    end

    def receive(msg_id, &forward)
      if msg_id
        synchronize(msg_id, &forward)
      else
        forward.call
      end
    end

    def synchronize(msg_id, &forward)
      with_finalizer do
        push([msg_id.to_i, forward])
        sort!
        while !empty? && last_id == self[0][0].to_i - 1
          @last_id = self[0][0].to_i
          shift[1].call
        end
      end
    end

    def with_finalizer
      EM.cancel_timer(finalizer)
      yield
      @finalizer = EM.add_timer(self.class.timeout) { finalize }
    end

    def finalize
      shift[1].call until empty?
      self.class.synchronizers.delete(id)
    end

    def sort!
      super { |lft, rgt| lft[0] <=> rgt[0] }
    end
  end
end

File: public/javascripts/lib/synchronizer.js
Travis.Synchronizer = function(forward) {
  this.forward  = forward;
  this.messages = [];
  this.lastId   = 0;
}
_.extend(Travis.Synchronizer, {
  synchronizers: {},
  receive: function(event, data, forward) {
    var buildId = data.build.id;
    this.synchronizers[buildId] = this.synchronizers[buildId] || new Travis.Synchronizer(forward);
    this.synchronizers[buildId].receive(event, data);
  }
});

_.extend(Travis.Synchronizer.prototype, {
  receive: function(event, data) {
    if(data.msg_id) {
      this.synchronize(event, data);
    } else {
      this.forward.apply(this, arguments);
    }
  },
  synchronize: function(event, data) {
    this.push(event, data);
    while(this.messages[0] && this.lastId == this.messages[0][1].msg_id - 1) {
      this.lastId = this.messages[0][1].msg_id;
      this.forward.apply(this, this.messages.shift());
    }
  },
  push: function() {
    this.messages.push(arguments);
    this.sort();
  },
  sort: function() {
    this.messages.sort(function(lft, rgt) { return lft[1].msg_id - rgt[1].msg_id; });
  },
});

Travis.receive = function(event, data) {
  // Travis.Synchronizer.receive(event, data, Travis.trigger);
  Travis.trigger(event, data);
};

